### PR TITLE
build: add conventional changelog support

### PR DESCRIPTION
**Pull request by tplooker**
This PR adds support for automated changelog generation during the release process

**Note** - this will enforce a specific commit format from now on so the changelog can be automatically generated based on the changelog.

Looking for feedback on this approach and confirming it is not going to be a developer burden.

### CODE REVIEW

**Comment by tplooker** [comment]
We will need to manually generate the changelog up until this date so later version releases will generate the changelog automatically

**Comment by OR13** [comment]
@tplooker what needs to be done to make this actionable?

**Comment by tplooker** [comment]
@OR13 first need to get the CI user access token sorted for the build pipeline, documented in #795, which is currently blocked as I'm awaiting user being setup in the npm org correctly

**Comment by OR13** [comment]
JFYI you can setup NPM Access tokens at an organization level in github...  https://github.blog/changelog/2020-05-14-organization-secrets/ @csuwildcat

File: commitlint.config.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+		'subject-case': [
+			2,
+			'never',
+			['sentence-case', 'start-case', 'pascal-case', 'upper-case']
+		],
+		"scope-empty": [2, "never"],
+        "scope-enum": [2, "always", [
+            "spec",
+            "ref-impl"
+        ]],
```
Comment (thehenrytsai) [inline commitlint.config.js:20]: Is this indented incorrectly?

File: commitlint.config.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+// Configuration is extended from @commitlint/config-conventional
+// @see https://github.com/conventional-changelog/commitlint/tree/master/@commitlint/config-conventional
+module.exports = {
+	rules: {
```
Comment (thehenrytsai) [inline commitlint.config.js:4]: This is probably a stretch ask. Is there a way to enforce an issue number to be associated with the commit message? like

#123 - fix(spec): ambiguity around update operation terminology

Comment (tplooker) [reply commitlint.config.js:4]: There are integrations like this but I would have to investigate a little more

File: commitlint.config.js (hunk 1 above)
Comment (tplooker) [reply commitlint.config.js]: Fixed now

File: commitlint.config.js (hunk 2 above)
Comment (OR13) [reply commitlint.config.js:4]: code reviews can address this. just reject the PR if it doesn't reference issues.

File: package.json [hunk 3]
```diff
@@ -33,29 +34,41 @@
...
     "lint": "tslint --fix --project .",
     "spec": "node ./scripts/spec-render.js",
-    "spec:edit": "node ./scripts/spec-edit.js"
+    "spec:edit": "node ./scripts/spec-edit.js",
+    "version": "conventional-changelog -p angular -i CHANGELOG.md -s && git add CHANGELOG.md"
   },
   "husky": {
     "hooks": {
-      "pre-commit": "npm run lint"
+      "pre-commit": "[[ -n $HUSKY_BYPASS ]] || npm run lint || pretty-quick --staged",
```
Comment (OR13) [inline package.json:65]: what is pretty-quick --staged doing?

Comment (thehenrytsai) [reply package.json:65]: I think automats the code in staged files.

Comment (tplooker) [reply package.json:65]: Good catch, nothing, will remove

Comment (thehenrytsai) [reply package.json:65]: Oh? You didn't intend to use `pretty-quick` for auto formatting?

Comment (tplooker) [reply package.json]: We can, but have to install and configure prettier too e.g see here https://github.com/mattrglobal/bls12381-key-pair/blob/master/package.json#L35

**Pull request merged**